src/App.js

```diff
@@ -1,25 +1,86 @@
-import logo from './logo.svg';
 import './App.css';
+import React, { useState } from 'react';
+
+export default function App() {
+  const [isValid, setIsValid] = useState(false);
+  const [message, setMessage] = useState('');
+
+  const emailRules = /\S+@\S+\.\S+/;
+  const passwordRules = /^(?=.*\d)(?=.*[!@#$%^&*])(?=.*[a-z])(?=.*[A-Z]).{8,}$/;
+
+  const validateEmail = (event) => {
+    const email = event.target.value;
+    if (emailRules.test(email)) {
+      setIsValid(true);
+      setMessage('✅ Your email looks good!');
+    } else {
+      setIsValid(false);
+      setMessage('❗Please enter a valid email!');
+    }
+  };
+
+  const validatePassword = (event) => {
```

Comment:
Same here. You could try using `useCallback` here.

Reply:
In a way, you should also be aware that not always does `useCallback` be useful. Because memoization comes with a cost. You can read more about this here: https://reactjs.org/docs/hooks-reference.html#usecallback. This would prevent input component from re-rendering unless the specific state changes. An eg would be to see what happens with and without usecallback to this function to the corresponding password input field when the email changes.

Also logically speaking, it would be better for this function to terminate as soon as an error case is found. Eg: if password length is less than 8, the message can be set and the function can return. This would prevent all other cases from being checked.